Update Træfik documentation to traefik:v2.x+ (#1118)
api-platform/docs: deployment/traefik.md

Then update the `docker-compose.yml` file belonging to your API Platform projects:

```diff

> # /anywhere/api-platform/docker-compose.yml

Reviewer: `anywhere`? We don't really use this elsewhere, do we?

Author: anywhere because it can be folder you want to store api-platform instance to but can be in the same directory than træfik instance

```

This way, you can configure your main variables into one single file.

> ## Multiple instances

Reviewer: ```suggestion
## Multiple Instances
```

> If you want to run multiple API Platform instances on same server, behind only one Træfik instance, you'll have to define different services names for each service to avoid named conflicts error since Træfik v2.0.  

Reviewer: ```suggestion
If you want to run multiple API Platform instances on the same server and behind only one Træfik instance, you'll have to define different service names for each service to avoid named conflicts error since Træfik v2.0.  
```

To achieve that, by setting only one more environment variable, you'll be able to make each instance unique. Here is a working example below:
```dotenv
# /anywhere/first/api-platform/.env
#...
RANDOM_UNIQUE_KEY=yourUniqueKeyForYourFirstInstance
```
```dotenv
# /anywhere/second/api-platform/.env
#...
RANDOM_UNIQUE_KEY=yourUniqueKeyForYourSecondInstance
```

Then update each traefik http routers names and services following this sample for admin
```yaml
# /anywhere/first/api-plaform/docker-compose.yml
# ...
    labels:
      - traefik.http.routers.admin-${RANDOM_UNIQUE_KEY}.rule=Host(`admin.${DOMAIN_NAME}`)
      - traefik.http.services.admin-${RANDOM_UNIQUE_KEY}.loadbalancer.server.port=3000
```

> ## More generic approach

Reviewer: ```suggestion
## More Generic Approach

```

Here is a fully working sample for Træfik generic config with a little script using docker-compose override approach.  
We assume that you've set `EXPOSE 3000` in your client and admin Dockerfile.

Create a new `init-dc.sh` which contains the generation code that will be written in `docker-compose.override.yaml` file.
```bash
#!/bin/sh

> # /anywhere/api-platform/init-dc.sh

Reviewer: See before.

> services=("admin:admin." "api:api." "cache:cache." "mercure:mercure." "ssr:") # Define your services keys following format "{container key}:{sub DNS}"

Reviewer: ```suggestion
services=("admin:admin." "api:api." "cache:cache." "mercure:mercure." "ssr:") # Define your services keys following this format: "{container key}:{sub DNS}"
```

text="version: '3.4'
services:"

for k in "${services[@]}" ; do
    key=${k%%:*}
    value=${k#*:}
    text+="
  $key:
    labels:
      - traefik.http.routers.$key.entrypoints=web-secure
      - traefik.http.routers.$key.rule=Host(\`${value}\${DOMAIN_NAME}\`)
      - traefik.http.routers.$key.tls=true
      - traefik.http.routers.$key.tls.domains[0].main=\${DOMAIN_NAME}
      - traefik.http.routers.$key.tls.domains[0].sans=*.\${DOMAIN_NAME}
      - traefik.http.routers.$key.tls.certresolver=sample
"
done

echo "$text" > ./docker-compose.override.yml
```

> Write this minimal configuration into your `traefik.toml` file

Reviewer: ```suggestion
Write this minimal configuration into your `traefik.toml` file:

```

```toml

> # /anywhere/traefik/traefik.toml

Reviewer: Same.

[providers.docker]
  endpoint = "unix:///var/run/docker.sock"

[api]
  insecure = true
  dashboard = true
  debug = true

[entryPoints.web]
  address = ":80"
  [entryPoints.web.http]
    [entryPoints.web.http.redirections]
      [entryPoints.web.http.redirections.entryPoint]
        to = "web-secure"
        scheme = "https"

[entryPoints.web-secure]
  address = ":443"
```

> Then after that update respectively your api-platform and Træfik `docker-compose.yml` following these example below.

Reviewer: ```suggestion
Then after that update respectively your API Platform and Træfik `docker-compose.yml` following these examples below:
```

    <<: *network

  admin:
    build:
      context: ./admin
      target: api_platform_admin_development
      cache_from:
        - ${ADMIN_IMAGE:-quay.io/api-platform/admin}
    image: ${ADMIN_IMAGE:-quay.io/api-platform/admin}
    tty: true
    depends_on:
      - dev-tls
    environment:
      # You should remove this variable from .env into client folder
      - REACT_APP_API_ENTRYPOINT=${HTTP_OR_SSL}api.${DOMAIN_NAME}
    volumes:
      - ./admin:/usr/src/admin:rw,cached
    <<: *network

volumes:
  db-data: {}
  dev-certs: {}

networks:
  api_platform_network:
    external: true
```

```yaml
# /anywhere/traefik/docker-compose.yml
version: '3.4'

x-network:
  &network
  networks:
    - api_platform_network

services:
  traefik:
    image: traefik:latest
    ports:
      - target: 80
        published: 80
        protocol: tcp
      - target: 443
        published: 443
        protocol: tcp
      - target: 8080
        published: 8080
        protocol: tcp

    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - ./traefik.toml:/etc/traefik/traefik.toml
      - ./acme.json:/acme.json
    <<: *network

networks:
  api_platform_network:
    external: true
```

> For more detailed step-by-step configuration, take a look at [this repository](https://github.com/darkweak/WorkshopContainous) which include Fail2ban link to Træfik instance.

Reviewer: ```suggestion
For a more detailed step-by-step configuration, take a look at [this repository](https://github.com/darkweak/WorkshopContainous) which include Fail2ban link to Træfik instance.
```
